0000-backtrace-in-core.md

```diff
@@ -0,0 +1,145 @@
+
+Currently, `Backtrace` is an essential part of the `std` library and user can control whether the backtrace is enabled or disabled using the environmental variable: `RUST_BACKTRACE`.
+
+In terms of Guide-level changes, there is not much to be discussed - only that it is moved to `core` and if `std` is not linked, automatic backtrace handlers will be generated. Otherwise, the regular implementation of `Backtrace` is present.
+
+TODO: what else should be here?
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following changes need to be made to implement this proposal:
+
+## Move the `Backtrace` to `core` and add a thin wrapper in `std`
+
+This way, a `StdBacktrace` struct is introduced in `std` that allows for `Debug` and `Display` formatting. 
+
+The regular API of `Backtrace` comprising `enabled()`, `create()` and `status()` would be left in the `std` as free-standing functions. Since, they are lang items they need to have a default implementation in case `std` is not linked, so they will be provided in such a form in the `core` library (and overwritten once `std` is linked):
+
+```rust
+/// Global implementation of backtrace functionality. Called to create
+/// `RawBacktrace` trait objects.
+#[cfg(not(bootstrap))]
+extern "Rust" {
+    #[lang = "backtrace_create"]
+    fn backtrace_create(ip: usize) -> *mut dyn RawBacktrace;
+
+    #[lang = "backtrace_enabled"]
+    fn backtrace_enabled() -> bool;
+
+    #[lang = "backtrace_status"]
+    fn backtrace_status(raw: *mut dyn RawBacktrace) -> BacktraceStatus;
+}
+
+#[cfg(bootstrap)]
+unsafe fn backtrace_create(_ip: usize) -> *mut dyn RawBacktrace {
+    UnsupportedBacktrace::create().inner
+}
+
+#[cfg(bootstrap)]
+unsafe fn backtrace_enabled() -> bool {
+    false
+}
+
+#[cfg(bootstrap)]
+unsafe fn backtrace_status(_raw: *mut dyn RawBacktrace) -> BacktraceStatus {
+    BacktraceStatus::Unsupported
+}
+
+```
+
+This change will make the `Backtrace` an optional part of the library and 
+This way, if the `Backtrace` is not enabled (on the library level) there is no need for it to report 
+
+Implementation-wise, `Backtrace` is **declared** in the `core` module and **defined** in `std` via *lang_items* which act as function hooks which are resolved during link-time. Special type `StdBacktrace` is introduced for the `std` part of implementation which acts as a proxy for `Debug` and `Display` trait impls.
+
+// TODO: add examples of how would one implement these functions themselves like panic hooks
+
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are actually many drawbacks, most important of them being `Backtrace` using a lot of `std` for the actual backtrace capturing. 
+
+The other one is a potential code bloat in `no_std` contexts, so a possible alternative may be only enabling the `Backtrace` conditionally via `cfg` settings. (not so sure about this though)
+
```

Comment:
Because we have a more abstract type we increase the size of 'pointer to Backtrace' from one word to two.

Reply:
You mean wrapping it in a `Box`?

Reply:
Whether it is a Box or a & type, pointers to trait objects are 'fat pointer's because we keep a pointer to the data and one to the vtable, not just a single pointer like we would for concrete types.